---
title: "16S rRNA Multi-Copy Cluster and Analysis"
author: "Nathan Olson and Ethan Ertel"
date: "December 7, 2015"
output: pdf_document
bibliography: bibliography.bib
---

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(magrittr)
library(stringr)
library(dplyr)
library(tidyr)
library(purrr)
library(readr)
library(ggplot2)
library(knitr)
library(ape)
```

## Abstract
Targeted genomic DNA amplification and sequencing of 16S rRNA is commonly used to characterize microbial communities in terms of identity and quantity. Current methods used to define community composition treat each sequence independently. However, microbial genomes may contain multiple copies of the 16S rRNA gene. Furthermore due to differential selective pressure, the gene sequence can vary between copies. The presence of both multiple different 16S sequences invalidates the assumption that clusters may be treated independently or used to uniquely identify species. Methods to correct for multiple copies only consider the gene copy number and not sequence diversity.  We analyzed 16S rRNA genes sequences from 2943 GenBank reference genomes and show that 16S genes from one organism may be represented by many clusters, and that the number of gene copies in a cluster can vary by strain. 
Include some kind of results - 5% of ambiguous clusters
As a initial attempt to develop a method that not only corrects for copy number but also taxonomic classification, we applied a simple set of linear constraints.  The linear constriants were unable to model the copy number mixed taxonomy clusters, we conclude the report with a discussion of other potential models that can be applied to address this issue.

## Introduction
Surveys of microbial communities aim to quantify the abundance of constituent microbes through the measurement of amplified segments of genomic DNA. Genes that produce ribosomal RNA, specifically the 16S rRNA sub-sequence, are favored targets for amplification and measurement because their high degree of conservation makes these genes an attractive basis for phylogenetic analysis and taxonomic identification. Thus, these genes are generally favored over other possible sequences despite disadvantages including, __NEED TO STATE DISADVANTAGES__  [@Vos2012]. In practice, measured sequences are clustered by similarity and these clusters are given a taxonomic identification. Edit distance, or the number of changes required to convert one sequence into another, is useful as a metric of pairwise similarity [@Ghodsi2011]. Clusters are commonly termed as operational taxonomic units (OTUs); OTUs are considered to be proxies for organismal abundance [@Wooley2010].

However, 16S rRNA sequences are known to be present in multiple copies [@Pei2010] and the presence of paralogous sequences presents the possibility that distantly related sequences may be identified as highly similar [@Koeppel2013]. 
__Issue with horizontal gene transfer__
A number of issues arise from this fact. Firstly, a single organism may be counted multiple times in many unique clusters. Secondly, paralogous (XXX) genes present in multiple species may cluster together, confounding identification; genes from species A may be labeled as species B if both map to a cluster labeled with the taxonomic information of species B.

__Potential diagram of copy number issue__

* Other studies that have characterized multi-copies
    * 
* Available methods correcting for gene copy number

+ Sequence matching
    * rnammer - how it works and what it does
+ Multiple sequence alignment
    * infernal - how it works and what it does
+ Phylogenetic Tree construction
    * Maximum likelihood tree
    * Evolutionary model
+ Sequence clustering
    * dnaClust - how it works and what is does

Need to read
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2432038/pdf/pone.0002566.pdf
http://bergelson.uchicago.edu/?p=886


## Methods
We downloaded 2943 microbial genomes and their taxonomic information from NCBI RefSeq (ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria) __(Benson et al. 2000)__. 16S rRNA gene sequences were extracted from the genomes using rnammer (http://www.cbs.dtu.dk/cgi-bin/sw_request?rnammer)[@Lagesen2007]. The extracted 16S rRNA genes were clustered with dnaclust [@Ghodsi2011]. THe sequences were clustered for a range of threshold values, where similarity scores for all pairwise comparisons of sequences within a cluster must be above the threshold value. Similarity score equals 1 - (edit distance)/length(shortest string), and ranges from 0 to 1. Typical threshold values used in 16S rRNA metagenomic studies range from 0.95 to 0.99; however, we investigate thresholds as low as 0.61 and as high as 1.00. 

See `multicopy_taxa_code.pdf` detailed methods including bash scripts and commands.

## Results
### Summary of Dataset
```{r echo=FALSE, message=FALSE}
cluster_df <- read_tsv("refseq_clusters/cluster_genome_count.tsv") %>% filter(!is.na(gi)) %>% 
    separate(cluster, c("clus","threshold","id"),  "_",remove = FALSE) %>% select(-clus)
```

```{r echo=FALSE, message=FALSE}
copy_count <- cluster_df %>%
    filter(threshold == "1.00") %>% 
    group_by(gi) %>% summarize(copies = n())
```

Number of 16S rRNA gene copies per genome ranged from `r min(copy_count$copies)` to `r max(copy_count$copies)` with a median of `r median(copy_count$copies)`.

```{r}
## Potentially drop plot, try to incorporate results into other figure
ggplot(copy_count) + geom_bar(aes(x = copies)) + 
    theme_bw() + labs(y = "Number of Genomes", x = "16S rRNA Gene Copies")
```

* 16S rRNA gene sequence diversity

```{r}
tree <- read.tree("infernal_raxml_tree/RAxML_result.infernal_16S_tree")
plot.phylo(tree)
tree_dist <- cophenetic.phylo(tree)
tree_dist_long <- tree_dist %>% as.data.frame() %>% 
    mutate(ID1 = rownames(tree_dist)) %>% gather(ID2, dist, -ID1)
```

```{r}
name_key <- read_tsv("name_key.tsv")

```

Next steps
Convert IDs to seq names
Address duplicate issues

Potentially filter outliers?

    * scatter/ boxplot
        * y-axis = within genome diversity /similarity
        * x-axis = taxa levels
        * use boxplots to show level diversity summary
        
### Clustering
The extracted 16S rRNA genes were clustered with dnaclust using a range of threshold values from 1.00 (identical gene sequences) to 0.61 (sequences with 61% similarity). The number of sequences assigned to a cluster was correlated with the clustering threshold.  For lower clustering (< 0.73) thresholds the majority of the sequences were assigned to a single cluster.

```{r echo = FALSE}
cluster_df_size <- cluster_df %>% group_by(threshold, id) %>% summarise(size = n()) 
```

```{r}
cor.test(x = cluster_df_size$size,y = as.numeric(cluster_df_size$threshold))
```

```{r}
ggplot(cluster_df_size) + geom_boxplot(aes(x = threshold, y = size)) + 
    scale_y_log10() + theme_bw() + 
    labs(x = "Clustering Threshold", y = "Sequences per Cluster")
```

Number of gene copies per genome by cluster
```{r}
cluster_copy_number <- cluster_df %>% group_by(threshold, id, gi, count) %>% summarise(size = n()) 
    # check size should equal count
#with(cluster_copy_number, sum(!(count == size)))
## checks out
cluster_copy_number %<>% select(-size) 
```

Variability in the number of 16S copies in a genome per cluster.  Variability in copy number within a cluster biases copy number correction when only a single value in copy number correction.  

```{r}
# cluster_copy_number %>% 
#     group_by(threshold, id) %>% 
#     summarize(number_sd =sd(count)) %>%
#     mutate(mixed_count = ifelse(number_sd == 0 | is.na(number_sd), "Uniform", "Mix")) %>% 
# ggplot() + geom_bar(aes(x = threshold, fill = mixed_count), position ="fill") + theme_bw()+ labs(fill = "Copy Number", x = "Clustering Threshold", y = "Proportion of Clusters")
```

Variability in copy number values
```{r echo = FALSE}
# cluster_copy_number %>% 
#     group_by(threshold, id) %>% 
#     summarize(number_sd =sd(count)) %>%
#     filter(number_sd > 0, !is.na(number_sd)) %>% 
# ggplot() + geom_jitter(aes(x = threshold,y = number_sd), position = position_jitter(width = 0.5)) + 
#     theme_bw() + labs(x = "Clustering Threshold", y = "Standard Deviation Copies per Genome")
```

```{r echo = FALSE}
# cluster_copy_number %>% 
#     group_by(threshold, id) %>% 
#     summarize(number_sd =sd(count)) %>%
#     mutate(mixed_count = ifelse(number_sd == 0 | is.na(number_sd), "Uniform", "Mix")) %>% 
#     #gather(pl_type, x, -threshold, -id) %>% 
#     filter(number_sd > 0 & !is.na(x), x != "Uniform") %>% 
#   ggplot() + geom_bar(aes(x = threshold, fill = mixed_count)) +geom_jitter(aes(x = threshold,y = number_sd), position = position_jitter(width = 0.5)) + theme_bw()+ labs(fill = "Copy Number", x = "Clustering Threshold", y = "Proportion of Clusters")  #+ facet_grid(pl_type~., scale = "free_y")
```

```{r echo = FALSE}
d1 <- cluster_copy_number %>% 
    group_by(threshold, id) %>% 
    summarize(number_sd =sd(count)) %>% 
    mutate(mixed_count = ifelse(number_sd == 0 | is.na(number_sd), 
                                "Uniform", "Mixed"))
d2 <- d1 %>% filter(number_sd > 0 & !is.na(number_sd))
d1$panel <- "Proportion"
d2$panel <- "Copy Number Standard Deviation"

d <- bind_rows(d1, d2)

p <- ggplot(data = d) + theme_bw()
p <- p + facet_wrap(~panel, ncol = 1, scale="free_y")
p <- p + geom_bar(data = d1, aes(x = threshold, fill = mixed_count), position = "fill")
p <- p + geom_jitter(data = d2, aes(x = threshold, y = number_sd), 
                     position = position_jitter(width = 0.5)) 
p <- p + labs( x = "Threshold", fill = "Cluster Type") + 
    theme(legend.position = "bottom", legend.direction = "horizontal", 
          strip.text = element_text(hjust = 0.02), axis.title.y = element_blank())
```

```{r echo = FALSE, fig.cap = "Variability number of 16S rRNA copies for sequences assigned to a cluster by genome. Top plot is the standard deviation of the copy numbers and bottom plot is the proportion of cluster with multiple copy numbers (Mixed), and single copy number values (Uniform)"}
p
```


Ambiguous clusters, clusters with more than one genome assigned
```{r echo=FALSE, warning=FALSE, message=FALSE}
cluster_df_ambig <- cluster_copy_number %>% group_by(threshold, gi) %>% 
    summarise(n_assigned = n()) %>% 
    mutate(ambig = ifelse(n_assigned > 1, "> 1","1"))
```

Proportion of genomes with 16S rRNA gene copies assigned to more than one cluster.
```{r}
cluster_df_ambig %>% 
    ggplot() +
        geom_bar(aes(x = threshold, fill = ambig)) + theme_bw() + 
        labs(x = "Clustering Threshold", y = "Number of Genomes", fill = "Assigned Clusters") +
    theme(legend.position = "bottom", legend.direction = "horizontal")
```

Number of clusters assigned, for genomes with more than one assigned gene copy.
```{r}
cluster_df_ambig %>% filter(n_assigned > 1, threshold != "1.00") %>% 
    ggplot() + 
        geom_bar(aes(x = threshold, fill = factor(n_assigned))) + 
        theme_bw() + labs(x = "Clustering Threshold", 
                          y = "Number of Genomes", 
                          fill = "Number of Assigned Clusters")
```

Taxonomic breakdown of clusters
```{r echo=FALSE, message=FALSE}
taxaID <- read_csv("clusters_with_taxonomic_annotation.csv") %>% filter(!is.na(gi))
cluster_df_anno <- taxaID %>% 
    separate(cluster, c("clus","threshold","id"),"_") %>% 
    select(-clus, -seq_names) %>% 
    group_by(threshold, id) %>% 
    gather(taxa_level, taxa, -threshold, -id, -seqs)

## Number of clusters with multiple taxa
cluster_df_anno_count <- cluster_df_anno %>% group_by(threshold, id, taxa_level) %>% 
    summarise(taxa_count = length(unique(taxa))) %>% 
    mutate(ambig = ifelse(taxa_count > 1, "multiple","single"))
#Threshold levels
thresholds <- cluster_df_size$threshold %>% unique() %>% as.numeric() %>% .[-15]
```

Taxonomic for 97% clustering threshold, >5% cluster classification error rate for genus and species level classifications.
```{r echo=FALSE, message=FALSE}
cluster_df_anno_count %>% group_by(threshold, taxa_level, ambig) %>% 
    summarise(ambig_count = n()) %>% group_by(threshold,taxa_level) %>%
    mutate(ambig_prop = ambig_count/sum(ambig_count)) %>% 
    filter(threshold != "1.00", 
               !(taxa_level %in% c("gi","taxid")),
               ambig == "multiple") %>% 
    ggplot() + geom_point(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
    geom_line(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
    scale_x_continuous(breaks = thresholds, 
                       labels = thresholds) +
    theme_bw() + labs(x = "Cluster Threshold", y = "Proportion of Clusters", color = "Taxonomic Level") +
    geom_hline(aes(yintercept = 0.05), linetype = 2, color = "grey40") + theme(legend.position = "bottom")
```

Combination of copy number and taxonomic ambiguity
```{r}
# ## Taxa ambiguous clusters assigned a value of 1 and non-ambiguous a value of 0
# cluster_taxa_ambig <- cluster_df_anno_count %>% 
#     group_by(threshold, taxa_level, ambig) %>% 
#     summarise(n_taxa = n()) %>% mutate(taxa_ambig = ifelse(n_taxa > 1, 1,0))
# 
# ## Copy number ambiguous clusters assigned a value of 2 and non-ambiguous a value of 0
# cluster_copy_ambig <- cluster_copy_number %>% group_by(threshold, gi) %>% 
#     summarise(n_copy = n()) %>% 
#     mutate(copy_ambig = ifelse(n_copy > 1, 2,0))
# 
# ## Cluster ambiguous for both ambig value of 3
# cluster_ambig <- full_join(cluster_taxa_ambig, cluster_copy_ambig) %>% 
#     rowwise() %>% mutate(ambig = copy_ambig + taxa_ambig)
```

```{r}
# cluster_ambig %>% group_by(threshold, taxa_level, ambig) %>% summarise(ambig_count = n()) %>% group_by(threshold,taxa_level,ambig) %>% mutate(ambig_prop = ambig_count/sum(ambig_count)) %>% 
#     filter(threshold != "1.00", 
#                !(taxa_level %in% c("gi","taxid")),
#                ambig == "multiple") %>% 
#     ggplot() + geom_point(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
#     geom_line(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
#     scale_x_continuous(breaks = thresholds, 
#                        labels = thresholds) +
#     theme_bw() + labs(x = "Cluster Threshold", y = "Proportion of Clusters", color = "Taxonomic Level") +
#     geom_hline(aes(yintercept = 0.05), linetype = 2, color = "grey40") + theme(legend.position = "bottom")
```

Want to try and generate a similar figure showing both clusters with both copy number ambiguity and taxonomic ambiguity.
```{r}
# cluster_taxa <- taxaID %>% 
#     separate(cluster, c("clus","threshold","id"),"_") %>% 
#     select(-clus, -seq_names) %>% left_join(cluster_copy_number)
# cluster_taxa_copy_ambig <- cluster_taxa %>% group_by(threshold, id) %>% 
#     mutate(unique_copy_num = count %>% length() %>% unique(),
#            ambig_copy = ifelse(unique_copy_num > 1, 2,0))
# cluster_taxa_copy_both_ambig <- cluster_taxa_copy_ambig %>% 
#     gather(taxa_level, taxa, -threshold, -id, -seqs, -unique_copy_num, -ambig_copy, -count) %>%  
#     group_by(threshold, id, taxa_level) %>% 
#     mutate(taxa_count = length(unique(taxa))) %>% 
#     mutate(ambig_taxa = ifelse(taxa_count > 1, 1,0))
# cluster_taxa_copy_both_ambig %>% group_by(threshold, id, count, taxa_level, ambig_copy, ambig_taxa) %>% 
#     summarise(ambig = n(), ambig_type = list(ambig_copy + ambig_taxa), ambig_length = length(ambig_type))
# #ambig_copy + ambig_taxa) %>% group_by(threshold, taxa_level, ambig) %>% summarise(count = n()) %>% group_by(threshold, taxa_level) %>% mutate(prop = count/sum(count))
```


Phylogenetic tree - showing relationship between clusters and taxa assignments



## Discussion
* Brief summary of findings  
* Comparison of copy number results to published results  
* When copy number correction can be applied  
* Issues - better methods for 16S extraction, CopyRighter    
* Limitations only using closed genomes - diversity  
    * 16Stimator method for getting copy number from draft assemblies, does not extract sequences
* Methods for correcting for copy number, primarily use single values, or proprtions for mixed copies, no methods to address taxonomic ambiguity - potential method for dealing with horizontal gene transfer.  


### When copy number correction does not work - ambiguous clusters
* Primer regions  
    * This work focuses on the whole 16S gene  
    * What are the expected results for single gene copies????  

### Recommendations
A number of approaches may use reference-based 16S multicopy information to guide taxonomic binning. If one assumes consistent and unbiased detection of 16S PCR products, the relationship between clustered OTUs 1..n and reference based counts of species 1..m is:
 Equation (XXX): A simple linear relationship between 16S rRNA copy and species count underlies the physical system
Organism identification could simply be treated as a linear optimization problem under these conditions if a suitable objective function c is defined. 

However, several major shortcomings hinder the success of this naive approach. PCR amplification is not uniform or linear, and so certain sequences will be overrepresented in such a way that the linear relationships no longer hold. In addition, the number of different OTUs detected is known to be a function of sequencing depth (Paulson et. al., 2013). Finally, the prediction of species counts is biased by the numbers and types of species present in the GenBank reference database. 

An alternative approach is to use 


Equation (XXX): The expected number of sequences in OTU i equals the sum of the weighted expectation for each component mapping to OTU i. These expectations depend on PDij (the probability of detection in a transcript from species j), number of copies of the relevant sequence, and statistical noise in the observation.

## References