---
title: "16S rRNA Multi-Copy Cluster and Analysis"
author: "Nathan Olson and Ethan Ertel"
date: "December 7, 2015"
output: 
  pdf_document: 
    fig_height: 4
    fig_width: 5
    fig_caption: yes
    latex_engine: xelatex
bibliography: bibliography.bib
---

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(magrittr)
library(stringr)
library(dplyr)
library(tidyr)
library(purrr)
library(readr)
library(ggplot2)
library(knitr)
library(kfigr)
library(ape)
library(ggtree)
library(colorspace)
opts_chunk$set(cache=TRUE)

## Getting reference info
## for(i in c("magrittr","stringr","dplyr","tidyr","purrr",
#            "readr","ggplot2","knitr","ape","ggtree", "colorspace")){
#     citation(i) %>% toBibtex() %>% print()
# }
```

## Abstract
Targeted genomic DNA amplification and sequencing of 16S rRNA is commonly used to characterize microbial communities in terms of identity and quantity. Current methods used to define community composition treat each sequence independently. However, microbial genomes may contain multiple copies of the 16S rRNA gene. Furthermore due to differential selective pressure, the gene sequence can vary between copies. The presence of both multiple different 16S sequences invalidates the assumption that clusters may be treated independently or used to uniquely identify species. Methods to correct for multiple copies only consider the gene copy number and not sequence diversity.  We analyzed 16S rRNA genes sequences from 2943 GenBank reference genomes and show that 16S genes from one organism may be represented by many clusters, and that the number of gene copies in a cluster can vary by strain. 
Include some kind of results - 5% of ambiguous clusters
As a initial attempt to develop a method that not only corrects for copy number but also taxonomic classification, we applied a simple set of linear constraints.  The linear constriants were unable to model the copy number mixed taxonomy clusters, we conclude the report with a discussion of other potential models that can be applied to address this issue.

## Introduction
Surveys of microbial communities aim to quantify the abundance of constituent microbes through the measurement of amplified segments of genomic DNA. Genes that produce ribosomal RNA, specifically the 16S rRNA sub-sequence, are favored targets for amplification and measurement because their high degree of conservation makes these genes an attractive basis for phylogenetic analysis and taxonomic identification. Thus, these genes are generally favored over other possible sequences despite disadvantages including, multiple gene copies within a genome and taxonomic ambiguity due to horizontal gene transfer [@Vos2012].

16S rRNA metagneomic datasets are commonly analyzed by clustering sequences based on a defined similarity. Edit distance, or the number of changes required to convert one sequence into another, is useful as a metric of pairwise similarity [@Ghodsi2011]. Clusters are defined as OTU, operational taxonomic units, and OTU counts are the number of sequences assigned to a cluster. Clusters are commonly termed as operational taxonomic units (OTUs); OTUs are considered to be proxies for organismal abundance [@Wooley2010]. Multiple 16S rRNA gene copies will bias the OTU counts as the number of sequences is not equivalent to the abundance of cells within the population being sampled assigned to the cluster [@Kembel2012]. Databases have been developed to characterize the number of gene copies per prokaryotic genome have been developed [@Stoddard2014; @Perisin2015], however these databases do not include the 16S rRNA gene copy sequences. A number of methods to normalize the OTU count values based on copy number estimates have been developed [@Kembel2012; @Angly2014; @Perisin2015]. All of these methods use a single copy number correction value for individual OTUs, therefore bacterial strains with different copy numbers assigned to a single OTU can bias the correction factor. The mutiple gene copy number and diversity of the gene copies within a genome either due to mutations or horizontal gene transfer [@Pei2010; @Koeppel2013] futher challenges copy number correction as assignment of gene copies to multiple clusters is not accounted for.  Additionally, due to horizontal gene transfer a 16S rRNA gene copy is a genome can be more similar to a 16S rRNA gene from a unrealted taxa than to other gene copies within the same genome.  No cluster taxonomic annotation methods take account for single genome multiple cluster assignment.

To address this issue we have developed a 16S rRNA gene copy sequence database.  The extent to whic h the number of gene copies per genome assigned to a cluster and number of clusters individual genomes were assigned was characterized using the gene copy sequence database.  We then attempted to develop a new method for copy number correction that accounts for the different numbers of assigned gene copies per cluster and genomes with multiple cluster assignments.  

## Methods
We downloaded 2943 microbial genomes and their taxonomic information from NCBI RefSeq (ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria) [@Pruitt2012]. 16S rRNA gene sequences were extracted from the genomes using rnammer (http://www.cbs.dtu.dk/cgi-bin/sw_request?rnammer)[@Lagesen2007]. The extracted 16S rRNA genes were clustered with dnaclust [@Ghodsi2011]. THe sequences were clustered for a range of threshold values, where similarity scores for all pairwise comparisons of sequences within a cluster must be above the threshold value. Similarity score equals 1 - (edit distance)/length(shortest string), and ranges from 0 to 1. Typical threshold values used in 16S rRNA metagenomic studies range from 0.95 to 0.99; however, we investigate thresholds as low as 0.61 and as high as 1.00. Exploratory data analysis was performed in R.

See `multicopy_taxa_code.pdf` detailed methods including bash scripts and commands.

## Results and Discussion
### 16S rRNA Multicopy Database
```{r echo=FALSE, message=FALSE}
cluster_df <- read_tsv("refseq_clusters/cluster_genome_count.tsv") %>% filter(!is.na(gi)) %>% 
    separate(cluster, c("clus","threshold","id"),  "_",remove = FALSE) %>% select(-clus)
```

```{r echo=FALSE, message=FALSE}
copy_count <- cluster_df %>%
    filter(threshold == "1.00") %>% 
    group_by(gi) %>% summarize(copies = n())
```

A database of 16S rRNA gene copy sequences was generated by extracting 16S rRNA gene sequences from all closed bacterial genomes in the NCBI RefSeq database.  Number of 16S rRNA gene copies per genome ranged from `r min(copy_count$copies)` to `r max(copy_count$copies)` with a median of `r median(copy_count$copies)` (`r figr(label = "copyDist", prefix = TRUE, link = TRUE, type = "Figure")`). The distribution of 16S rRNA gene copy numbers agrees with previous studies [@Angly2014; @Vetrovsky2013], exluding a genome with 16S rRNA gene copies. Only a single method was used to identify 16S rRNA gene copy sequences and therefore may contain sequences that were incorrectly identified as 16S rRNA genes, as indicated by outliers in the database phylogenetic tree (`r figr(label = "fullTree", prefix = TRUE, link = TRUE, type = "Figure")`). Therefore the results below are presented as a proof of concept. A more rigorus approach to generating the sequence copy database similar to that used in Angly et al. (2014) would help validate the database. 

```{r copyDist, anchor = "Figure", fig.align="center", echo=FALSE, message=FALSE, fig.cap="Distribution in the number of gene copies present in a bacterial genome.", fig.height=3,fig.width=3}
## Potentially drop plot, try to incorporate results into other figure
ggplot(copy_count) + geom_bar(aes(x = copies)) + 
    theme_bw() + labs(y = "Number of Genomes", x = "16S rRNA Gene Copies")
```

```{r fullTree, anchor = "Figure", fig.align="center", echo=FALSE, message=FALSE, fig.cap="Phylogeny of 16S rRNA gene copy database used in this study. Long branch length indicate potential sequences in the database that were incorrectly identified as 16S rRNA genes by rnammer.", fig.height=6,fig.width=6}
tree <- read.tree("infernal_raxml_tree/RAxML_result.infernal_16S_tree")
ggtree(tree, layout="circular")
```

### Clustering
The extracted 16S rRNA genes were clustered with dnaclust using a range of threshold values from 1.00 (identical gene sequences) to 0.61 (sequences with 61% similarity). For lower clustering (< 0.73) thresholds the majority of the sequences were assigned to a single cluster (`r figr(label = "clusterSize", prefix = TRUE, link = TRUE, type = "Figure")`).

```{r echo = FALSE}
cluster_df_size <- cluster_df %>% group_by(threshold, id) %>% summarise(size = n()) 
```

```{r clusterSize, anchor = "figure", fig.align="center", echo=FALSE, message=FALSE, fig.cap="Distribution of cluster size, number of sequences assigned to a cluster, by clustering threshold.", fig.height=4,fig.width=5}
ggplot(cluster_df_size) + geom_boxplot(aes(x = threshold, y = size)) + 
    scale_y_log10() + theme_bw() + 
    labs(x = "Clustering Threshold", y = "Sequences per Cluster")
```

### Cluster Copy Number 
The number of 16S rRNA gene copies from a single genome within a cluster varies. The variabily in copy number within a cluster biases copy number correction when only a single value in copy number correction (`r figr(label = "copyNumVar", prefix = TRUE, link = TRUE, type = "Figure")`).  Single cluster copy number value is currently the only method used in copy number correction [@Kembel2012; @Angly2014; @Stoddard2014; @Perisin2015].  CopyRighter [@Angly2014], attempts to address this issue by defining the cluster copy number correction value proportionally based on the copy number of the species assigned to a cluster.  

```{r echo = FALSE}
cluster_copy_number <- cluster_df %>% group_by(threshold, id, gi, count) %>% summarise(size = n()) 
    # check size should equal count
#with(cluster_copy_number, sum(!(count == size)))
## checks out
cluster_copy_number %<>% select(-size) 

d1 <- cluster_copy_number %>% 
    group_by(threshold, id) %>% 
    summarize(number_sd =sd(count)) %>% 
    mutate(mixed_count = ifelse(number_sd == 0 | is.na(number_sd), 
                                "Uniform", "Mixed"))
d2 <- d1 %>% filter(number_sd > 0 & !is.na(number_sd))
d1$panel <- "Proportion"
d2$panel <- "Copy Number Standard Deviation"

d <- bind_rows(d1, d2)

p <- ggplot(data = d) + theme_bw()
p <- p + facet_wrap(~panel, ncol = 1, scale="free_y")
p <- p + geom_bar(data = d1, aes(x = threshold, fill = mixed_count), position = "fill")
p <- p + geom_jitter(data = d2, aes(x = threshold, y = number_sd), 
                     position = position_jitter(width = 0.5)) 
p <- p + labs( x = "Threshold", fill = "Cluster Type") + 
    theme(legend.position = "bottom", legend.direction = "horizontal", 
          strip.text = element_text(hjust = 0.02), axis.title.y = element_blank())
```

```{r copyNumVar, anchor="Figure", fig.align="center",echo = FALSE, fig.cap = "Variability number of 16S rRNA copies for sequences assigned to a cluster by genome. Top plot is the standard deviation of the copy numbers and bottom plot is the proportion of cluster with multiple copy numbers (Mixed), and single copy number values (Uniform)", fig.height=6,fig.width=6}
p
```

Assignment of 16S rRNA gene sequences from a single genome to multiple clusters further confound the use of single copy nuber correction values. While 16S rRNA gene copies from a single genome are assigned to only one cluster, a number of genomes have 16S rRNA gene sequences assigned to multiple clusters (`r figr(label = "ambigAssign", prefix = TRUE, link = TRUE, type = "Figure")`).  Lowering the clustering threshold has little effect on the number of clusters copies of the 16S rRNA gene is assigned to for a genome. The database used in this analysis was generated using only a single method for 16S rRNA gene extraction.  Visualization of the database phylogeny indicates that a number of outlier sequences are included in the database which are likely responsible for a few of the multiple cluster genome sequence copy assignments, especially thoes with mutiple assigned clusters when using lower threshold values.

```{r echo=FALSE, warning=FALSE, message=FALSE}
cluster_df_ambig <- cluster_copy_number %>% group_by(threshold, gi) %>% 
    summarise(n_assigned = n()) %>% 
    mutate(ambig = ifelse(n_assigned > 1, "> 1","1"))
cluster_df_ambig_prop <- cluster_df_ambig %>% 
    group_by(threshold, n_assigned) %>% summarise(count = n()) %>%
    group_by(threshold) %>% mutate(prop_n_assigned = count/sum(count), 
                                   check = sum(prop_n_assigned))
```

```{r ambigAssign, anchor="Figure", fig.align="center",echo=FALSE, message=FALSE, fig.cap="Proportion of genomes assigned to multiple clusters by clustering threshold.", fig.height=4,fig.width=6}
cluster_df_ambig_prop %>% 
    filter(threshold %in% c("0.70", "0.85", "0.91", "0.94", "0.97", "0.99")) %>% 
    ggplot() + 
        geom_point(aes(x = threshold, y = count, color = factor(n_assigned), 
                       fill = factor(n_assigned)), 
                   position = position_jitterdodge()) + 
        theme_bw() + labs(x = "Clustering Threshold", 
                          y = "Proportion of Genomes", 
                          color = "Number of Assigned Clusters") +
        scale_fill_discrete(guide = "none") +
        scale_y_log10(breaks = c(1, 5, 10, 50, 100, 500, 1000, 3000)) +
        theme(legend.position = "bottom")
```



### Cluster Taxonomic Ambiguity

```{r echo=FALSE, message=FALSE}
taxaID <- read_csv("clusters_with_taxonomic_annotation.csv") %>% filter(!is.na(gi))
cluster_df_anno <- taxaID %>% 
    separate(cluster, c("clus","threshold","id"),"_") %>% 
    select(-clus, -seq_names) %>% 
    group_by(threshold, id) %>% 
    gather(taxa_level, taxa, -threshold, -id, -seqs)

## Number of clusters with multiple taxa
cluster_df_anno_count <- cluster_df_anno %>% group_by(threshold, id, taxa_level) %>% 
    summarise(taxa_count = length(unique(taxa))) %>% 
    mutate(ambig = ifelse(taxa_count > 1, "multiple","single"))
#Threshold levels
thresholds <- cluster_df_size$threshold %>% unique() %>% as.numeric() %>% .[-15]

anno_count_fig_df <- cluster_df_anno_count %>% group_by(threshold, taxa_level, ambig) %>% 
    summarise(ambig_count = n()) %>% group_by(threshold,taxa_level) %>%
    mutate(ambig_prop = ambig_count/sum(ambig_count)) %>% 
    filter(threshold != "1.00", 
               !(taxa_level %in% c("gi","taxid")),
               ambig == "multiple")
min_ambig_lt97 <- anno_count_fig_df %>% ungroup() %>% 
    mutate(threshold = as.numeric(threshold)) %>% 
    filter(threshold < 0.97) %>% .$ambig_prop %>% min() %>% round(2)
```

Taxonomic ambiguity for 97% clustering threshold, >5% cluster classification error rate for genus and species level classifications (`r figr(label = "taxaAmbig", prefix = TRUE, link = TRUE, type = "Figure")`). For clustering thresholds less than 0.97, the proportion of taxonomically ambiguous clusters is greater than `r min_ambig_lt97`. When analyzing 16S rRNA metagenomic datasets, the presence of taxonomically ambiguous clusters must be taken into consideration. 

```{r taxaAmbig, anchor="Figure", fig.align="center",echo=FALSE, message=FALSE, fig.cap="Proportion of cluster with more than one assigned taxa by clustering threshold and taxonomic level.", , fig.height=4,fig.width=6}
anno_count_fig_df %>%  
    ggplot() + geom_point(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
    geom_line(aes(x = as.numeric(threshold), y = ambig_prop, color = taxa_level)) +
    scale_x_continuous(breaks = thresholds, 
                       labels = thresholds) +
    theme_bw() + labs(x = "Cluster Threshold", 
                      y = "Proportion of Clusters", color = "Taxonomic Level") +
    geom_hline(aes(yintercept = 0.05), linetype = 2, color = "grey40") + 
    theme(legend.position = "bottom")
```

### Phylogenetic Analysis

```{r echo=FALSE, message=FALSE}
cluster_df_anno_labled <- cluster_df_anno %>% group_by(threshold, id, taxa_level) %>% 
    mutate(taxa_count = length(unique(taxa)),
           ambig = ifelse(taxa_count > 1, "multiple","single"),
           cluster = paste0("cluster_",threshold,"_",id)) %>% 
    filter(threshold == "0.97", taxa_level == "genus")
    
name_key <- read_tsv("name_key.tsv") %>% select(seqs, rep_name)
cluster_97_name_key <- cluster_df_anno_labled %>% left_join(name_key)
tree2 <- drop.tip(tree, c("ID.2237","ID.2946","ID.1312","ID.1795","ID.891",
                          "ID.2838","ID.569","ID.1459", "ID.4043"))
ambig_node_labels <- cluster_97_name_key %>% filter(ambig == "multiple") %>% 
    .$rep_name %>% unique()
non_ambig_node_labels <- cluster_97_name_key %>% filter(ambig != "multiple") %>% 
    .$rep_name %>% unique()
```

The phylogenetic distribution of ambiguous clusters (`r figr(label = "ambigTree", prefix = TRUE, link = TRUE, type = "Figure")`). When clustering at the 0.97 threshold and using genus level taxonomic annotation, sequences belonging to clusters with multiple assigned taxa are distributed throughout the phylogenetic tree. The wide distribution of sequences assigned to ambiguous clusters indicates that ambiguity in cluster taxonomic annotation is challenges taxonomic classification issues for a majority of taxonomic lineages.

```{r ambigTree, anchor="Figure", fig.align="center",fig.align="center",echo=FALSE, message=FALSE, fig.cap="16S rRNA gene phylogeny, teal braches indicate 16S rRNA gene sequences assigned to taxonomically ambiguous clusters, at the 0.97 clustering threshold and genus level taxonomic assignment. To provide better branch resolution in the phyogenetic tree, potential outlier 16S rRNA sequences were exlucded from the phylogenetic tree below", fig.height=6,fig.width=6}
#plot(tree)
cols <- rainbow_hcl(2)
tree3 <- groupOTU(tree2, focus=ambig_node_labels)
ggtree(tree3, layout="circular", aes(color=group)) + scale_color_manual(values = cols, breaks=1:2, labels=c("Ambiguous","Defined"))+ 
    theme(legend.position="bottom")
```

### Resolving Copy Number and Taxonomic Cluster Ambiguity

A number of approaches may use reference-based 16S multicopy information to guide taxonomic binning. If one assumes consistent and unbiased detection of 16S PCR products, the relationship between clustered OTUs 1..n and reference based counts of species 1..m is:
 Equation (XXX): A simple linear relationship between 16S rRNA copy and species count underlies the physical system
Organism identification could simply be treated as a linear optimization problem under these conditions if a suitable objective function c is defined. 

However, several major shortcomings hinder the success of this naive approach. PCR amplification is not uniform or linear, and so certain sequences will be overrepresented in such a way that the linear relationships no longer hold. In addition, the number of different OTUs detected is known to be a function of sequencing depth [@Paulson2013]. Finally, the prediction of species counts is biased by the numbers and types of species present in the GenBank reference database. 

An alternative approach is to use 


Equation (XXX): The expected number of sequences in OTU i equals the sum of the weighted expectation for each component mapping to OTU i. These expectations depend on PDij (the probability of detection in a transcript from species j), number of copies of the relevant sequence, and statistical noise in the observation.

## References